python/pyiceberg/expressions/base.py

```diff
@@ -342,235 +243,321 @@ def __str__(self) -> str:
 
 
 @dataclass(frozen=True)
-class AlwaysTrue(BooleanExpression, ABC, Singleton):
+class AlwaysTrue(BooleanExpression, Singleton):
     """TRUE expression"""
 
     def __invert__(self) -> AlwaysFalse:
         return AlwaysFalse()
 
 
 @dataclass(frozen=True)
-class AlwaysFalse(BooleanExpression, ABC, Singleton):
+class AlwaysFalse(BooleanExpression, Singleton):
     """FALSE expression"""
 
     def __invert__(self) -> AlwaysTrue:
         return AlwaysTrue()
 
 
-class IsNull(UnboundPredicate[T]):
-    def __invert__(self) -> NotNull:
-        return NotNull(self.term)
+@dataclass(frozen=True)
+class BoundPredicate(Bound[T], BooleanExpression):
+    term: BoundTerm[T]
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals is not None:
-            raise AttributeError("Null is a unary predicate and takes no Literals.")
+    def __invert__(self) -> BoundPredicate[T]:
+        raise NotImplementedError
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundIsNull[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundIsNull(bound_ref)
 
+@dataclass(frozen=True)
+class UnboundPredicate(Unbound[T, BooleanExpression], BooleanExpression):
+    as_bound: ClassVar[type]
+    term: UnboundTerm[T]
 
-class BoundIsNull(BoundPredicate[T]):
-    def __invert__(self) -> BoundNotNull:
-        return BoundNotNull(self.term)
+    def __invert__(self) -> UnboundPredicate[T]:
+        raise NotImplementedError
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("Null is a unary predicate and takes no Literals.")
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
+        raise NotImplementedError
 
 
-class NotNull(UnboundPredicate[T]):
-    def __invert__(self) -> IsNull:
-        return IsNull(self.term)
+@dataclass(frozen=True)
+class UnaryPredicate(UnboundPredicate[T]):
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
```

Comment:
Should this return type be `BoundUnaryPredicate`? Same question for `SetPredicate.bind(...)` and `LiteralPredicate.bind(...)` which can have return types of `BoundSetPredicate` and `BoundLiteralPredicate`.

Reply:
Binding may produce a different type of expression. For example, if `IsNull` is bound to a required type, binding will return `AlwaysFalse`